experimental/ir/option.go

```diff
@@ -0,0 +1,308 @@
+// *Options descriptor at the root is not stored.
+func (o Option) Field() Field {
+	return wrapField(o.Context(), o.raw.field)
+}
+
+// Value returns the value this option is set to.
+func (o Option) Value() Value {
+	return wrapValue(o.Context(), o.raw.value)
+}
+
+func wrapOption(c *Context, p arena.Pointer[rawOption]) Option {
+	if c == nil || p.Nil() {
+		return Option{}
+	}
+	return Option{
+		withContext: internal.NewWith(c),
+		raw:         c.arenas.options.Deref(p),
+	}
+}
+
+// Value is an evaluated expression assigned to an [Option].
+type Value struct {
+	withContext
+	raw *rawValue
+}
+
+type rawValue struct {
+	ast  ast.ExprAny
+	bits rawValueBits
+	ty   predeclared.Name
+
+	isArray                  bool // If set, value is an arena.Pointer[[]rawValueBits].
+	isUninterpretedPath      bool // If set, value is an intern.ID.
+	isUninterpretedAggregate bool // If set, value is an intern.ID.
+}
+
+// rawValueBits is used to represent the actual value for all types, according to
+// the following encoding:
+//  1. All numeric types, including bool and enums. This holds the bits.
+//  2. String and bytes. This holds an intern.ID.
+//  3. Messages. This holds an arena.Pointer[rawMessage].
+//  4. Repeated fields. This holds an arena.Pointer[[]rawValueBits], where each
+//     value is interpreted as a non-array with this value's type.
+//     This exploits the fact that arrays cannot contain other arrays.
+type rawValueBits uint64
+
+// AST returns the expression that evaluated to this value.
+func (v Value) AST() ast.ExprAny {
+	return v.raw.ast
+}
+
+// Type is the element type of this value. For arrays, this is the type of the
+// element; Protobuf does not have a concept of an "array type".
+func (v Value) Type() Type {
+	elems := v.Elements()
+	if elems.Len() == 0 {
+		return Type{}
+	}
+	return elems.At(0).Type()
+}
+
+// Elements returns an indexer over the elements within this value.
+//
+// If the value is not an array, it contains the singular element within;
+// otherwise, it returns the elements of the array.
+//
+// If this is an uninterpreted option, returns an empty indexer.
+func (v Value) Elements() seq.Indexer[Element] {
+	var slice []rawValueBits
+	switch {
+	case v.raw.isUninterpretedPath || v.raw.isUninterpretedAggregate:
+		slice = nil
+	case v.raw.isArray:
+		slice = *v.Context().arenas.arrays.Deref(arena.Pointer[[]rawValueBits](v.raw.bits))
+	default:
+		slice = slicesx.One(&v.raw.bits)
+	}
+
+	return seq.Slice[Element, rawValueBits]{
+		Slice: slice,
+		Wrap: func(bits *rawValueBits) Element {
+			return Element{v.withContext, *bits, v.raw.ty}
+		},
+	}
+}
+
+// Element is an element within a [Value].
+//
+// This exists because array values contain multiple non-array elements; this
+// type provides uniform access to such elements.
+type Element struct {
+	withContext
+	bits rawValueBits
+	// Unknown means this is a message type.
+	// Enum types are predeclared.Int32.
+	ty predeclared.Name
```

Comment:
These seems unfortunate, that you can't inspect the enum type of an enum value. Maybe instead encode enums/messages/groups as negative values here and map to a separate enum for `NamedType`?

Reply:
Well, one thing that you could do is say that `predeclared.Unknown` represents "is a user type", and determining whether it's a message or enum requires looking at the type.